// rxjs@7.8.2/src/internal/testing/ColdObservable.ts
import { Observable } from '.pnpm/rxjs@7.8.2/node_modules/rxjs/src/internal/Observable';
import { Subscription } from '.pnpm/rxjs@7.8.2/node_modules/rxjs/src/internal/Subscription';
import { Scheduler } from '.pnpm/rxjs@7.8.2/node_modules/rxjs/src/internal/Scheduler';
import { TestMessage } from '.pnpm/rxjs@7.8.2/node_modules/rxjs/src/internal/testing/TestMessage';
import { SubscriptionLog } from '.pnpm/rxjs@7.8.2/node_modules/rxjs/src/internal/testing/SubscriptionLog';
import { SubscriptionLoggable } from '.pnpm/rxjs@7.8.2/node_modules/rxjs/src/internal/testing/SubscriptionLoggable';
import { applyMixins } from '.pnpm/rxjs@7.8.2/node_modules/rxjs/src/internal/util/applyMixins';
import { Subscriber } from '.pnpm/rxjs@7.8.2/node_modules/rxjs/src/internal/Subscriber';
import { observeNotification } from '.pnpm/rxjs@7.8.2/node_modules/rxjs/src/internal/Notification';

export class ColdObservable<T> extends Observable<T> implements SubscriptionLoggable {
  public subscriptions: SubscriptionLog[] = [];
  scheduler: Scheduler;
  // @ts-ignore: Property has no initializer and is not definitely assigned
  logSubscribedFrame: () => number;
  // @ts-ignore: Property has no initializer and is not definitely assigned
  logUnsubscribedFrame: (index: number) => void;

  constructor(public messages: TestMessage[], scheduler: Scheduler) {
    super(function (this: Observable<T>, subscriber: Subscriber<any>) {
      const observable: ColdObservable<T> = this as any;
      const index = observable.logSubscribedFrame();
      const subscription = new Subscription();
      subscription.add(
        new Subscription(() => {
          observable.logUnsubscribedFrame(index);
        })
      );
      observable.scheduleMessages(subscriber);
      return subscription;
    });
    this.scheduler = scheduler;
  }

  scheduleMessages(subscriber: Subscriber<any>) {
    const messagesLength = this.messages.length;
    for (let i = 0; i < messagesLength; i++) {
      const message = this.messages[i];
      subscriber.add(
        this.scheduler.schedule(
          (state) => {
            const { message: { notification }, subscriber: destination } = state!;
            observeNotification(notification, destination);
          },
          message.frame,
          { message, subscriber }
        )
      );
    }
  }
}
applyMixins(ColdObservable, [SubscriptionLoggable]);
